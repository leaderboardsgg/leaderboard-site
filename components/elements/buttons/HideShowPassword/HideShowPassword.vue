<script setup lang="ts">
import BaseButton from 'elements/buttons/BaseButton/BaseButton.vue'

interface HideShowPasswordState {
  hidden: boolean
}

const state: HideShowPasswordState = reactive({
  hidden: true,
})

const hiddenRef = toRef(state, 'hidden')

function toggleHidden() {
  hiddenRef.value = !hiddenRef.value
}
</script>

<template>
  <BaseButton
    id="hide-show-password"
    data-testid="hide-show-password-button"
    class="button"
    v-bind="$attrs"
    @click="toggleHidden"
    @keyup.enter="toggleHidden"
  >
    <i-svg-eye-hidden
      v-show="hiddenRef"
      class="h-5 w-5"
      data-testid="hidden-eye-icon"
    />
    <i-svg-eye-visible
      v-show="!hiddenRef"
      class="h-5 w-5"
      data-testid="visible-eye-icon"
    />
  </BaseButton>
</template>

<style lang="postcss" scoped>
.button {
  @apply hidden sm:inline bg-white;
}
</style>
