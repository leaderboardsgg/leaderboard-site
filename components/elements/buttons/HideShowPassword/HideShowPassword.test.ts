import { describe, expect, test } from 'vitest'
import { mount } from '@vue/test-utils'

import HideShowPassword from './HideShowPassword.vue'

describe('<HideShowPassword />', () => {
  const HideShowPasswordWrapper = mount(HideShowPassword)

  test('should render without crashing', () => {
    expect(HideShowPasswordWrapper.isVisible()).toBe(true)
  })

  const passwordButton = HideShowPasswordWrapper.get(
    '[data-testid="hide-show-password-button"]',
  )
  const hiddenEyeIcon = HideShowPasswordWrapper.get(
    '[data-testid="hidden-eye-icon"]',
  )
  const visibleEyeIcon = HideShowPasswordWrapper.get(
    '[data-testid="visible-eye-icon"]',
  )
  describe('Visible eye, hidden eye, and the button', () => {
    test('should render without crashing', () => {
      expect(passwordButton.isVisible()).toBe(true)
      expect(hiddenEyeIcon.isVisible()).toBe(true)
      expect(visibleEyeIcon.isVisible()).toBe(true)
    })
  })

  describe('when clicking the button', () => {
    test('should toggle the state', async () => {
      expect(hiddenEyeIcon.html()).not.toContain('style="display: none;')
      expect(visibleEyeIcon.html()).toContain('style="display: none;')

      await passwordButton.trigger('click')

      expect(hiddenEyeIcon.html()).toContain('style="display: none;')
      expect(visibleEyeIcon.html()).not.toContain('style="display: none;')
    })
  })

  //  this contains the previous flip so we need to do the opposite
  describe('when the button is focus and the enter key up event is triggered', () => {
    test('should toggle the state', async () => {
      expect(hiddenEyeIcon.html()).toContain('style="display: none;')
      expect(visibleEyeIcon.html()).not.toContain('style="display: none;')

      await passwordButton.trigger('keyup.enter')

      expect(hiddenEyeIcon.html()).not.toContain('style="display: none;')
      expect(visibleEyeIcon.html()).toContain('style="display: none;')
    })
  })
})
