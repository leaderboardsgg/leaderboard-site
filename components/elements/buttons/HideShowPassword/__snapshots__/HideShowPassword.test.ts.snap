// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HideShowPassword /> renders correctly 1`] = `
<button
  class="button button"
  data-testid="hide-show-password-button"
  id="hide-show-password"
>
  <svg-icon-stub
    class="h-5 w-5"
    data-testid="hidden-eye-icon"
    name="eye-hidden"
  />
   
  <svg-icon-stub
    class="h-5 w-5"
    data-testid="visible-eye-icon"
    name="eye-visible"
    style="display: none;"
  />
</button>
`;
